pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -162,6 +165,9 @@ protected abstract void addOrReplaceSegment(ImmutableSegmentImpl segment, Thread
       Iterator<RecordInfo> recordInfoIterator, @Nullable IndexSegment oldSegment,
       @Nullable MutableRoaringBitmap validDocIdsForOldSegment);
 
+  protected abstract void removeExpiredPrimaryKeys(IndexSegment segment, Iterator<RecordInfo> recordInfoIterator,
```

Comment:
Can you add some javadocs for this new method? 

Also, I think this method should be included as a part of the `PartitionUpsertMetadataManager` interface itself since it manipulates the partition metadata. Is there a reason why it is not included as a part of the interface?

Reply:
@navina this method is similar to `addOrReplaceSegment()` method that is being called internally by replaceSegment.